### ISSUE TITLE
curly: option to allow omission when there is only one statement *and* it is on the same line

### ISSUE THREAD

**Issue by hgwood** [eslint/eslint#1812]
Rationale: retains the advantages of the `curly` rule, while allowing compactness when the condition and its statement are trivial and not important enough to take up 3 lines (which may amount to 75% of the code of the function it's in).

Valid
```
if (foo) doSomething();
else doSomethingElse();

if (foo) {
  doSomething();
} else {
  doSomethingElse();
}
```

Invalid
```
if (foo)
  doSomething();
else
  doSomethingElse();

if (foo
  && bar) baz();

if (foo) foo(
  bar,
  baz);
```

Does it make sense? I'm willing to take a shot at coding it.

**Comment by nzakas** [comment]
Fine by me. I'd want to be sure the option name for this is clear.

**Comment by hgwood** [comment]
What about "allow-one-liners"?

**Comment by nzakas** [comment]
Hmm, I don't find that to be very clear. We also haven't used "allow" in this rule at all, so that doesn't fit.  Maybe something with the word "collapsed"? Because this isn't just about single lines, it's about a collapsed expression.

**Comment by hgwood** [comment]
I agree that "allow" doesn't feel ideal. I'm trying to come up with a name that is coherent with what already exists:
- `curly` alone means "there should always be curly braces present"
- `curly` with `multi` means "there should be curly braces present *if and only if* the block contains multiple statements"
- `curly` with the new option means "there should always be curly braces *unless* the whole statement is on the same line"

While `multi` hardens the rule by adding a condition for the presence of braces, the new option relaxes it. This is why I suggested "allow".

Another way to do it would be to make the new option mean "there should be curly braces *if and only if* the whole statement spans multiple lines". This implies that `if (foo) { return; }` is invalid. However this spawns 2 new problems:
- it conflicts with `allowSingleLine` from `brace-style`
- the logical name would be...`multi`, but this time meaning multiple lines in the whole statement instead of multiple statement in the block.
I therefore suggest we don't do this.

What we could do is copy from `brace-style` and have `allowSingleLine` like this: `"curly": [2, {"allowSingleLine": true}]`.

Now on the subject of line vs collapsed, I do think it is about single lines. For example, I wouldn't want the rule to accept the following statements, even though they could be thought as collapsed:

```js
if (foo
  && bar
  && baz) return;

if (foo) bar(
  baz,
  baz);
```

These statements are already 3 lines long so they must be important, and them spanning 2 more lines would be OK, recommended even. The new option is meant to allow *trivial* statements only. Trivial meaning "so simple it's nearly irrelevant".

What do you think?

**Comment by nzakas** [comment]
How about `multi-line`? Seems like a nice add-on to `multi`

**Comment by hgwood** [comment]
Let's go with that. :)

**Comment by khoomeister** [comment]
Just to clarify, can we have braces with a single statement in `multi-line`?

For example, if both the `if` condition and statement are really long, then it helps to break them up into multiple lines.  Example:
```javascript
if (a fairly long conditional) {
  doSomething('some really long string', etc);
}
```

**Comment by hgwood** [comment]
Yes, absolutely. `multi-line` will not warn about such code. It will not warn either with `if (foo) { bar(); }` (that's the job of `brace-style`). `multi-line` only allows to be brace-less in some situations, it does not enforce brace-lessness.

**Comment by khoomeister** [comment]
Awesome thanks @hgwood :smiley:
